dashboard/test_combined_data.py

```diff
@@ -0,0 +1,209 @@
+@patch('combined_data.boto3.client')
+def test_download_csv_exception(mock_client, mock_getcwd, mock_readcsv, mock_remove, caplog):
+    """Test case when something goes wrong in the downloading of csv from the S3 or the conversion to dataframe, an error is raised."""
+
+    mock_s3 = MagicMock()
+    mock_client.return_value = mock_s3
+    mock_client.download_file.side_effect = Exception()
+
+    bucket_name = "test-bucket"
+    folder_name = "test-folder"
+    file_name = "temp_file.csv"
+    with caplog.at_level(logging.INFO):
+        df = download_csv_from_s3_to_dataframe(
+            bucket_name, folder_name, file_name)
+
+    assert df is None
+    assert "An error occurred:" in caplog.text
+
+
+@patch('combined_data.pd.read_sql_query')
+@patch('combined_data.get_connection')
+def test_fetch_keyword_recordings_success(mock_conn, mock_read_sql, aws_env_vars):
+    """Test successful return of dataframe version of keyword recording table."""
+    mock_s3 = MagicMock()
+    mock_conn.return_value = mock_s3
+
+    mock_read_sql.return_value = pd.DataFrame({
+        'keyword_recordings_id': [1],
+        'keywords_id': [10],
+        'total_mentions': [100],
+        'avg_sentiment': [0.5],
+        'date_and_hour': ['2024-12-12 12:00:00']
+    })
+
+    df = fetch_keyword_recordings_as_dataframe()
+
+    assert isinstance(df, pd.DataFrame)
+    assert not df.empty
+    mock_conn.assert_called_once()
+    mock_read_sql.assert_called_once()
+    mock_s3.close.assert_called_once()
+
+
+@patch('combined_data.pd.read_sql_query')
+@patch('combined_data.get_connection')
+def test_fetch_keyword_recordings_error(mock_conn, mock_read_sql, aws_env_vars):
+    """Test fetching keywords returns an error if needed."""
+    mock_s3 = MagicMock()
+    mock_conn.return_value = mock_s3
+    mock_conn.side_effect = psycopg2.Error
+
+    df = fetch_keyword_recordings_as_dataframe()
+
+    assert df is None
+    mock_conn.assert_called_once()
+    mock_read_sql.assert_not_called()
+
+
+@patch("combined_data.fetch_keyword_recordings_as_dataframe")
+@patch("combined_data.download_csv_from_s3_to_dataframe")
+@patch('combined_data.get_connection')
+def test_main_combine_success(mock_conn, mock_download, mock_fetch_df, aws_env_vars, caplog):
+    """Test the main combine function successful downloads a csv file and returns a dataframe."""
+    mock_download.return_value = pd.DataFrame({
+        'keyword_recordings_id': [1],
+        'keywords_id': [10],
+        'total_mentions': [100],
+        'avg_sentiment': [0.5],
+        'date_and_hour': ['2024-12-12 12:00:00']
+    })
+    mock_fetch_df.return_value = pd.DataFrame({
+        'keyword_recordings_id': [2],
+        'keywords_id': [20],
+        'total_mentions': [10],
+        'avg_sentiment': [-.1],
+        'date_and_hour': ['2024-12-12 06:00:00']
+    })
+
+    with caplog.at_level(logging.INFO):
+        result = main_combine()
+
+    assert isinstance(result, pd.DataFrame)
+    assert not result.empty
```

Comment:
we could put these asserts together 
assert isinstance(result, pd.DataFrame) and not result.empty